### ISSUE TITLE
[Feature Request] Add a method to treat dependency charts as if they were libraries

### ISSUE THREAD

**Issue by philomory** [helm/helm#8533]
TL;DR: We should have a template function similar to `template` for rendering entire subcharts with specific values from within chart code, i.e.

```
{{- subchart "my-subchart" (dict "some_url" (printf "https://foo.%s.example.com" .Values.environment)) }}
```
# The Problem
One of the problems I butt up against fairly regularly is the awkwardness of composing Charts (in a so-called "umbrella chart"). Specifically, if you have an umbrella chart which contains several dependencies, the only way to encode the values they have in common is to write out those values multiple times, once for each chart:

```yaml
subchart1:
  some_key: foo
  another_key: bar

subchart2:
  some_key: foo
  another_key: bar
```

# Options that don't quite cut it
Of course, there _is_ the `globals` construct, so you could do something like:

```yaml
globals:
  some_key: foo
  another_key: bar
```

This option has a couple of flaws; first of all, it only works if the child charts actually know they're supposed to read from `.Values.globals` rather than from `.Values`; most public charts don't do that. Second, it only works in cases similar to above, where the different subcharts actually have the same names for their keys. It wouldn't work, for example, as a means of consolidating the data below, even though the _values_ are the same in each subchart:

```yaml
subchart1:
  some_key: foo
  another_key: bar

subchart2:
  some_param: foo
  another_param: bar
```

The globals option also doesn't work if the subchart values are similar but not identical, as in this example:

```yaml
subchart1:
  some_key: foo1
  another_key: bar1

subchart2:
  some_key: foo2
  another_key: bar2
```

What I'd really like is a way to _derive_ values for the subcharts from the values of the umbrella chart, and then pass those derived values as "arguments" to the subcharts. In the past, people have asked for templating of `values.yaml` as a way to achieve this, a request that's been rejected (and whose rejection I pretty much agree with).

A different way to accomplish this would be to make your subcharts into "helper charts" whose component templates get included from the umbrella chart; then, your values file could look like:

```yaml
some_thing: foo
another_thing: bar
```

... and your umbrella chart could have a template like ...

```
{{- $values1 := dict }}
{{- $_ := set $values1 "some_key" (printf "%s%d" .Values.some_thing 1) }}
{{- $_ := set $values1 "another_key" (printf "%s%d" .Values.another_thing 1) }}
{{- $scope1 := dict "Values" $values1 "Release" .Release }}
{{- template "subchart1.template" $scope1 }}

{{- $values2 := dict }}
{{- $_ := set $values2 "some_param" (printf "%s%d" .Values.some_thing 2) }}
{{- $_ := set $values2 "another_param" (printf "%s%d" .Values.another_thing 2) }}
{{- $scope2 := dict "Values" $values2 "Release" .Release }}
{{- template "subchart2.template" $scope2 }}
```

Now, this is something you could do _now_, but it has three problems of its own:

1. The subcharts can't be regular charts anymore; if you took a subchart that was structured this way and tried to deploy it, it wouldn't do anything because all the templates are partials/helpers. In any case this means you can't actually use this technique to encapsulate public charts without forking them and altering them fairly significantly.
2. The `template` function only takes a single template, not the whole subchart; this means you have to know how many templates are in your subcharts and what all their names are.
3. You lose access to the default values defined within the subcharts; you can't even manually merge them in, because parent charts don't normally have access to values from subcharts. You could use the `import-values` feature in the dependencies list, but to merge in _all_ of the defaults from the subchart, you'd have to enumerate every single key for which a default value is defined (in fact, you'd have to enumerate them twice: once in the dependencies list, and once in your template to make a dict to merge in).

# The proposed solution
My proposal is to add a flag for dependencies that would sit alongside other keys such as `enabled`, `condition`, and `import-values`, called `library`; the default value of this field would be `false` (alternatively, the field could be called `mode` and take `library` as one possible value, alongside `subchart` (the default) and potentially other values in the future).

If a dependency is specified as a `library`, then the templates it contains will _not_ be rendered by default; however, unlike a subchart which is simply `disabled`, a `library` subchart could by dynamically rendered, using a new template function:

The new template function, `subchart`, would work very similarly to the `template` function that already exists, in that it would render templates defined elsewhere. However, it would have somewhat different behavior:

* The first argument is the name of a subchart, not of a template.
* The second argument is mandatory rather than optional, and must be a dict
* The second argument does not represent the entire scope for the rendering of the subchart; instead, it is treated as if it were the contents of an "additional values" file passed to the `--values` command line flag; that is, it is merged with the default values from inside the subchart, and the merged result is available during subchart rendering as `.Values`. 

Final usage would end up looking like this:

```
{{- $values1 := dict }}
{{- $_ := set $values1 "some_key" (printf "%s%d" .Values.some_thing 1) }}
{{- $_ := set $values1 "another_key" (printf "%s%d" .Values.another_thing 1) }}
{{- subchart "subchart1" $values1 }}

{{- $values2 := dict }}
{{- $_ := set $values2 "some_param" (printf "%s%d" .Values.some_thing 2) }}
{{- $_ := set $values2 "another_param" (printf "%s%d" .Values.another_thing 2) }}
{{- subchart "subchart2" $values2 }}
```

Whether `.Release`, `.Files`, etc. should be copied verbatim from the parent chart or be passed as arguments is open to discussion.

# Pros and cons
The advantages of such a setup would be:

* It allows for more sophisticated ways of composing charts into larger charts
* It allows dynamically deriving values you'll pass to subcharts
* It doesn't require modifications to the subcharts
* It doesn't require adding templating support to `values.yaml`
* It doesn't interfere with the existing behavior of subcharts (i.e. it is not a breaking change)

Arguably, a disadvantage would be that the proposal is "complicated", however, I don't think it's actually that bad:
* From an implementation standpoint, it mostly draws on functionality already present in Helm like merging explicit and default values, and rendering charts to text.
* From a usage standpoint it's a _little_ bit complicated, but it's still less complicated than currently available alternatives for accomplishing the same goal (such as forking subcharts to make them use `globals`) while also being more flexible.

**Comment by bacongobbler** [comment]
Hi @philomory. I haven't had a chance to look at the full proposal here yet, but I have heard of a similar use case before. Have you had a chance to look at other proposals like https://github.com/helm/helm/issues/3242 to handle your use case?

PR https://github.com/helm/helm/pull/7477 implements that proposal, should you be interested to try that out and see if that will work for you.

**Comment by philomory** [comment]
@bacongobbler Thanks for the pointers, I thought I had dug around in existing suggestions, but obviously I didn't dig far enough to find #3242. That said, from reading it I'm not entirely confident the proposal in #3242 is sufficient for our needs (I'll have to think about it). I do feel like I want the ability to _derive_ subchart values _as a function of_ parent chart values. That derivation has to be done in code, and right now the only place code can go in a Helm charts is in the templates.

**Comment by philomory** [comment]
OK, actually having re-read #3242, it's more functional than I first gave it credit for. It's still doesn't provide as much flexibility as I was hoping for, and I think `Chart.yaml` is kind of a weird place for this sort of functionality, but, given that there's already a (mostly?) working implementation I'll definitely give it a shot.

**Comment by hickeyma** [comment]
@philomory If this is aligned with #3242 then it might be worth reaching out to @withnale and see if a solution can be agreed on. PR #7477 is yet to be verified and it would be good to find a solution to the feature. If this makes sense then it might be worth closing this proposal as a duplicate of #3242?

**Comment by philomory** [comment]
@hiceyma While the problem this proposal is trying to solve is (roughly) the same as the problem in #3242, the actual proposed solution is different enough that I'd prefer to keep this open until I've had a chance to mull it over a bit more. Of course, if the Helm maintainers strongly believe that the solution proposed in #3242 is their preferred option, then by all means we should close this issue, but, I'm not sure it's strictly a _duplicate_ per se.

Unless you're suggesting I migrate my proposed solution into a comment on #3242?

**Comment by hickeyma** [comment]
@philomory I am ok as you look to work this out. I just want to avoid 2 solutions for the same use case.
